placement/algo/algo.go

```diff
@@ -0,0 +1,482 @@
+	if isHostAbsent(s, leavingHost) {
+		return nil, nil, errHostAbsent
+	}
+	var leavingHostShards *hostShards
+	var hosts []*hostShards
+	for _, phs := range s.HostShards() {
+		h := newHostShards(phs)
+		if phs.Host().Address == leavingHost.Address {
+			leavingHostShards = h
+			continue
+		}
+		hosts = append(hosts, h)
+	}
+	ps := newPlacement(hosts, s.Shards(), s.Replicas())
+	return newPlaceShardingHelper(ps, s.Replicas(), true), leavingHostShards, nil
+}
+
+func newPlaceShardingHelper(ps placementSnapshot, targetRF int, hostCapacityAscending bool) *placementHelper {
+	ph := &placementHelper{
+		shardToHostMap: make(map[uint32]map[*hostShards]struct{}),
+		rackToHostsMap: make(map[string]map[*hostShards]struct{}),
+		rf:             targetRF,
+		hostShards:     ps.hostShards,
+		uniqueShards:   ps.Shards(),
+	}
+
+	// build rackToHost map
+	ph.buildRackToHostMap()
+
+	// build targetLoad map for hostHeap
+	ph.buildHostHeap(hostCapacityAscending)
+	return ph
+}
+
+func (ph *placementHelper) buildRackToHostMap() {
+	for _, h := range ph.hostShards {
+		if _, exist := ph.rackToHostsMap[h.hostRack()]; !exist {
+			ph.rackToHostsMap[h.hostRack()] = make(map[*hostShards]struct{})
+		}
+		ph.rackToHostsMap[h.hostRack()][h] = struct{}{}
+		for shard := range h.shardsSet {
+			ph.assignShardToHost(shard, h)
+		}
+	}
+}
+
+func (ph *placementHelper) buildHostHeap(hostCapacityAscending bool) {
+	overSizedRack := 0
+	overSizedHosts := 0
+	totalHostNumber := len(ph.hostShards)
+	rackSizeMap := make(map[string]int)
+	for rack := range ph.rackToHostsMap {
+		rackHostNumber := len(ph.rackToHostsMap[rack])
+		if float64(rackHostNumber)/float64(totalHostNumber) >= 1.0/float64(ph.rf) {
+			overSizedRack++
+			overSizedHosts += rackHostNumber
+		}
+		rackSizeMap[rack] = rackHostNumber
+	}
+
+	targetLoad := make(map[string]int)
+	for _, host := range ph.hostShards {
+		rackSize := rackSizeMap[host.hostRack()]
+		if float64(rackSize)/float64(totalHostNumber) >= 1.0/float64(ph.rf) {
+			// if the host is on a over-sized rack, the target load is topped at shardLen / rackSize
+			targetLoad[host.hostAddress()] = int(math.Ceil(float64(ph.getShardLen()) / float64(rackSize)))
+		} else {
+			// if the host is on a normal rack, get the target load with aware of other over-sized rack
+			targetLoad[host.hostAddress()] = ph.getShardLen() * (ph.rf - overSizedRack) / (totalHostNumber - overSizedHosts)
+		}
+	}
+
+	ph.hostHeap = newHostHeap(ph.hostShards, hostCapacityAscending, targetLoad, ph.rackToHostsMap)
+}
+
+func (ph placementHelper) getHostLen() int {
+	return len(ph.hostShards)
+}
+
+func (ph placementHelper) getShardLen() int {
+	return len(ph.uniqueShards)
+}
+
+func (ph placementHelper) getAvgLoad() int {
+	totalLoad := ph.rf * ph.getShardLen()
+	numberOfHosts := ph.getHostLen()
+	return totalLoad / numberOfHosts
+}
+
+func (ph placementHelper) getShardSet(shards []uint32) map[uint32]struct{} {
```

Comment:
This doesn't seem  to depend on `ph`, make it a function instead of a method perhaps?